Marlin/src/inc/Conditionals_LCD.h

```diff
@@ -675,6 +675,163 @@
+      #if NONE(FIL_RUNOUT3_PULLUP, FIL_RUNOUT3_PULLDOWN)
+        #define FIL_RUNOUT3_PULLDOWN
+      #endif
+      #if NONE(FIL_RUNOUT4_PULLUP, FIL_RUNOUT4_PULLDOWN)
+        #define FIL_RUNOUT4_PULLDOWN
+      #endif
+      #if NONE(FIL_RUNOUT5_PULLUP, FIL_RUNOUT5_PULLDOWN)
+        #define FIL_RUNOUT5_PULLDOWN
+      #endif
+      #if NONE(FIL_RUNOUT6_PULLUP, FIL_RUNOUT6_PULLDOWN)
+        #define FIL_RUNOUT6_PULLDOWN
+      #endif
+      #if NONE(FIL_RUNOUT7_PULLUP, FIL_RUNOUT7_PULLDOWN)
+        #define FIL_RUNOUT7_PULLDOWN
+      #endif
+      #if NONE(FIL_RUNOUT8_PULLUP, FIL_RUNOUT8_PULLDOWN)
+        #define FIL_RUNOUT8_PULLDOWN
+      #endif
+    #endif // ENABLED(FIL_RUNOUT_PULLUP)
+  #endif // EITHER(FIL_RUNOUT_PULLUP, FIL_RUNOUT_PULLDOWN)
+#else // !ENABLED(DISTINCT_FIL_RUNOUT_STATES)
+  #undef FIL_RUNOUT1_STATE
+  #undef FIL_RUNOUT1_PULLUP
+  #undef FIL_RUNOUT1_PULLDOWN
+  #define FIL_RUNOUT1_STATE FIL_RUNOUT_STATE
+  #define FIL_RUNOUT1_PULLUP FIL_RUNOUT_PULLUP
+  #define FIL_RUNOUT1_PULLDOWN FIL_RUNOUT_PULLDOWN
+  #undef FIL_RUNOUT2_STATE
+  #undef FIL_RUNOUT2_PULLUP
+  #undef FIL_RUNOUT2_PULLDOWN
+  #if NUM_RUNOUT_SENSORS > 1
+    #define FIL_RUNOUT2_STATE FIL_RUNOUT_STATE
+    #define FIL_RUNOUT2_PULLUP FIL_RUNOUT_PULLUP
+    #define FIL_RUNOUT2_PULLDOWN FIL_RUNOUT_PULLDOWN
+  #endif
+  #undef FIL_RUNOUT3_STATE
+  #undef FIL_RUNOUT3_PULLUP
+  #undef FIL_RUNOUT3_PULLDOWN
+  #if NUM_RUNOUT_SENSORS > 2
+    #define FIL_RUNOUT3_STATE FIL_RUNOUT_STATE
+    #define FIL_RUNOUT3_PULLUP FIL_RUNOUT_PULLUP
+    #define FIL_RUNOUT3_PULLDOWN FIL_RUNOUT_PULLDOWN
+  #endif
+  #undef FIL_RUNOUT4_STATE
+  #undef FIL_RUNOUT4_PULLUP
+  #undef FIL_RUNOUT4_PULLDOWN
+  #if NUM_RUNOUT_SENSORS > 3
+    #define FIL_RUNOUT4_STATE FIL_RUNOUT_STATE
+    #define FIL_RUNOUT4_PULLUP FIL_RUNOUT_PULLUP
+    #define FIL_RUNOUT4_PULLDOWN FIL_RUNOUT_PULLDOWN
+  #endif
+  #undef FIL_RUNOUT5_STATE
+  #undef FIL_RUNOUT5_PULLUP
+  #undef FIL_RUNOUT5_PULLDOWN
+  #if NUM_RUNOUT_SENSORS > 4
+    #define FIL_RUNOUT5_STATE FIL_RUNOUT_STATE
+    #define FIL_RUNOUT5_PULLUP FIL_RUNOUT_PULLUP
+    #define FIL_RUNOUT5_PULLDOWN FIL_RUNOUT_PULLDOWN
+  #endif
+  #undef FIL_RUNOUT6_STATE
+  #undef FIL_RUNOUT6_PULLUP
+  #undef FIL_RUNOUT6_PULLDOWN
+  #if NUM_RUNOUT_SENSORS > 5
+    #define FIL_RUNOUT6_STATE FIL_RUNOUT_STATE
+    #define FIL_RUNOUT6_PULLUP FIL_RUNOUT_PULLUP
+    #define FIL_RUNOUT6_PULLDOWN FIL_RUNOUT_PULLDOWN
+  #endif
+  #undef FIL_RUNOUT7_STATE
+  #undef FIL_RUNOUT7_PULLUP
+  #undef FIL_RUNOUT7_PULLDOWN
+  #if NUM_RUNOUT_SENSORS > 6
+    #define FIL_RUNOUT7_STATE FIL_RUNOUT_STATE
+    #define FIL_RUNOUT7_PULLUP FIL_RUNOUT_PULLUP
+    #define FIL_RUNOUT7_PULLDOWN FIL_RUNOUT_PULLDOWN
+  #endif
+  #undef FIL_RUNOU8_STATE
+  #undef FIL_RUNOU8_PULLUP
+  #undef FIL_RUNOU8_PULLDOWN
+  #if NUM_RUNOUT_SENSORS > 7
+    #define FIL_RUNOU8_STATE FIL_RUNOUT_STATE
+    #define FIL_RUNOU8_PULLUP FIL_RUNOUT_PULLUP
+    #define FIL_RUNOU8_PULLDOWN FIL_RUNOUT_PULLDOWN
+  #endif
```

Comment:
These are all missing the T in FIL_RUNOUT8_STATE

Reply:
I guess that is an argument for adding that 8-extruder test case!

Reply:
Oh wow, that's a good eye!

I've added the tests to `BIGTREE_GTR_V1_0-tests`, but I wonder if they would had caught this?

Do you know if Marlin has any plans to added more comprehensive tests? I think there isn't much that we could test eg test that `FIL_RUNOUT8_STATE ` is correctly set, or to test that if we have a particular configuration that it would fail.

Reply:
I don't know whether the compilation-only test would have caught this or not.

I am not aware of current plans to add more extensive testing. It is something that has been in the back of my mind and I've chatted with others about a little.

Retroactively adding tests to large existing codebases is excruciatingly difficult. Unit tests for new code would be nice, but becomes complicated due to Marlin's dependence on conditional compilation and global variables.

I _personally_ think it is going to be essential to figure something out, but the path forward isn't very clear.

(My opinions are my own, and don't actually represent the project)

Reply:
Have you figured out how to run tests locally on your machine prior to pushing to the PR?

It probably takes a little bit of path work, but you can run them using the following command. I do it on Windows inside a Windows Subsystem for Linux shell.
`run-tests . BIGTREE_GTR_V1_0`

Reply:
Running tests locally is probably not exactly documented anywhere...
Maybe somebody will surprise me with a link!

Reply:
* Running locally: I've been trying to do that but I had no luck with PlatformIO in WSL as I get an SSL Error: `Error: HTTPSConnectionPool(host='api.registry.ns1.platformio.org', port=443): Max retries exceeded with url: /v3/packages?query=name%3A%22ststm32%22+type%3A%22platform%22 (Caused by SSLError(CertificateError("hostname 'api.registry.ns1.platformio.org' doesn't match either of '*.platformio.org', 'platformio.org'",),))`
* Introducing tests to a big codebase: I think the best approach would be to, as you said, set up the testing infrastructure, and adding code for new modules, and slowly introduce tests for the existing base as you find regressions/issues
* I would be interested in helping with a testing approach, if that's something you'd be thinking about. I might even try to see how would that look like, but unfortunately I have very little knowledge/experience with C++

Reply:
I'm setting up a docker-compose workflow for testing locally, so hopefully I won't have any quirks finding out how to run tests.

Reply:
I managed to have a running docker-compose setup locally: #19981 

Although it's quite slower than CI (but I think because the code is in a Windows drive, not inside WSL)